mainchain-follower/db-sync-follower/src/candidates/datum/registered.rs

```diff
@@ -55,82 +56,64 @@ pub struct AdaBasedStaking {
 	pub signature: MainchainSignature,
 }
 
-impl TryFrom<&Datum> for RegisterValidatorDatum {
+impl TryFrom<PlutusData> for RegisterValidatorDatum {
 	type Error = super::Error;
 
-	fn try_from(datum: &Datum) -> super::Result<Self> {
+	fn try_from(datum: PlutusData) -> super::Result<Self> {
 		decode_legacy_register_validator_datum(datum).ok_or("Invalid registration datum".into())
 	}
 }
 
-pub fn decode_legacy_register_validator_datum(datum: &Datum) -> Option<RegisterValidatorDatum> {
-	match datum {
-		ConstructorDatum { constructor: 0, fields } => {
-			let stake_ownership = fields.first().and_then(decode_ada_based_staking_datum)?;
-			let sidechain_pub_key = fields
-				.get(1)
-				.and_then(|d| d.as_bytestring())
-				.map(|bytes| SidechainPublicKey(bytes.clone()))?;
-			let sidechain_signature = fields
-				.get(2)
-				.and_then(|d| d.as_bytestring())
-				.map(|bytes| SidechainSignature(bytes.clone()))?;
-			let consumed_input = fields.get(3).and_then(decode_utxo_id_datum)?;
-			let _own_pkh = fields.get(4).and_then(|d| d.as_bytestring())?;
-			let aura_pub_key = fields
-				.get(5)
-				.and_then(|d| d.as_bytestring())
-				.map(|bytes| AuraPublicKey(bytes.clone()))?;
-			let grandpa_pub_key = fields
-				.get(6)
-				.and_then(|d| d.as_bytestring())
-				.map(|bytes| GrandpaPublicKey(bytes.clone()))?;
-			Some(RegisterValidatorDatum::V0 {
-				stake_ownership,
-				sidechain_pub_key,
-				sidechain_signature,
-				consumed_input,
-				aura_pub_key,
-				grandpa_pub_key,
-			})
-		},
-
-		_ => None,
-	}
+pub fn decode_legacy_register_validator_datum(datum: PlutusData) -> Option<RegisterValidatorDatum> {
+	let fields = datum
+		.as_constr_plutus_data()
+		.filter(|datum| datum.alternative().is_zero())
+		.filter(|datum| datum.data().len() >= 7)?
+		.data();
+	let stake_ownership = decode_ada_based_staking_datum(fields.get(0))?;
+	let sidechain_pub_key = fields.get(1).as_bytes().map(SidechainPublicKey)?;
+	let sidechain_signature = fields.get(2).as_bytes().map(SidechainSignature)?;
+	let consumed_input = decode_utxo_id_datum(fields.get(3))?;
+	let _own_pkh = fields.get(4).as_bytes()?;
+	let aura_pub_key = fields.get(5).as_bytes().map(AuraPublicKey)?;
+	let grandpa_pub_key = fields.get(6).as_bytes().map(GrandpaPublicKey)?;
+	Some(RegisterValidatorDatum::V0 {
+		stake_ownership,
+		sidechain_pub_key,
+		sidechain_signature,
+		consumed_input,
+		aura_pub_key,
+		grandpa_pub_key,
+	})
 }
 
-fn decode_ada_based_staking_datum(datum: &Datum) -> Option<AdaBasedStaking> {
-	match datum {
-		ConstructorDatum { constructor: 0, fields } => match fields.first().zip(fields.get(1)) {
-			Some((ByteStringDatum(f0), ByteStringDatum(f1))) => {
-				let pub_key = TryFrom::try_from(f0.clone()).ok()?;
-				Some(AdaBasedStaking { pub_key, signature: MainchainSignature(f1.clone()) })
-			},
-			_ => None,
-		},
-		_ => None,
-	}
+fn decode_ada_based_staking_datum(datum: PlutusData) -> Option<AdaBasedStaking> {
```

Comment:
The filtering is almost a bit too explicit. Personally I would probably prefer a more explicit early exit for this one:

```Rust
fn decode_ada_based_staking_datum(datum: PlutusData) -> Option<AdaBasedStaking> {
    let constr = datum.as_constr_plutus_data()?;
    if constr.alternative() != 0 || constr.data().len() < 2 {
        return None;
    }

    let fields = constr.data();
    let pub_key = TryFrom::try_from(fields.get(0).as_bytes()?).ok()?;
    let signature = MainchainSignature(fields.get(1).as_bytes()?);

    Some(AdaBasedStaking { pub_key, signature })
}
```

Reply:
But there are many ways to skin the cat and I think if we cannot keep previously approved code (because PlutusData API is worse), then I'm easy how you do it. I don't see a bug here.

Reply:
Filtering is more readable and allows us to use positive statements about the data instead of their negations.